File: server/utils/transaction.py
import json

from utils.sign import verifySign

from db.account import getUserBalance, getUserByCpf, updateUserBalance
from db.transaction import addTransaction, getAllUserTransactions


def executeTransaction(request):
    message = "transaction executed"
    statusCode = 200
    transactionStatus = "validada"

    # verificar se a assinatura é válida
    if not verifySign(request):
        message = "invalid sign"
        statusCode = 401
        transactionStatus = "assinatura invalida"

    transaction_hex = request.transaction
    transaction_bytes = bytes.fromhex(transaction_hex)
    transaction_str = transaction_bytes.decode("utf-8")
    transaction_dict = json.loads(transaction_str)

    # verificar saldo do sender
    senderBalance = getUserBalance(transaction_dict["sender"])
    if senderBalance < transaction_dict["amount"]:
        message = "insufficient balance"
        statusCode = 400
        transactionStatus = "saldo insuficiente"

    # verificar se o recipient existe
    recipient = getUserByCpf(transaction_dict["recipient"])
    if recipient is None:
        message = "recipient not found"
        statusCode = 404
        transactionStatus = "destinatario invalido"
        transaction_dict["recipient"] = None

    # atualizar saldo do sender e do recipient
    if statusCode == 200:
        updateUserBalance(transaction_dict["sender"], -transaction_dict["amount"])
        updateUserBalance(transaction_dict["recipient"], transaction_dict["amount"])

    # salvar transacao no banco
    addTransaction(
        transaction_dict["sender"],
        transaction_dict["recipient"],
        transaction_dict["amount"],
        transactionStatus,
    )

    return message, statusCode


def executeGetUserTransactions(cpf: str):
    transactions = getAllUserTransactions(cpf)
    mappedTransactions = []
    for transaction in transactions:
        mappedTransaction = {
            "id": transaction[0],
            "cpf_sender": transaction[1],
            "name_sender": transaction[2],
            "cpf_recipient": transaction[3],
            "name_recipient": transaction[4],
            "amount": float(transaction[5]),
            "event_datetime": transaction[6].isoformat(),
            "status": transaction[7],
        }
        mappedTransactions.append(mappedTransaction)
    return mappedTransactions

File: view/src/utils/transaction.ts
import { format, parseISO } from "date-fns";

import { signData } from "@/utils/rsa";
import { get, post } from "./api";

export interface Transaction {
  sender: string;
  recipient: string;
  amount: number;
  date: string;
}

interface SignedTransaction {
  transaction: Transaction;
  sign: Buffer;
}

interface PayloadTransaction {
  transaction: string;
  sign: string;
}

type TransactionStatus =
  | "validada"
  | "saldo insuficiente"
  | "assinatura invalida"
  | "destinatario invalido"
  | "erro nao mapeado";

interface FetchedTransaction {
  id: number;
  cpf_sender: string;
  name_sender: string;
  cpf_recipient: string;
  name_recipient: string;
  amount: number;
  event_datetime: string;
  status: TransactionStatus;
}

export interface MappedTransaction {
  transactionId: number;
  senderCpf: string;
  senderName: string;
  recipientCpf: string;
  recipientName: string;
  amount: number;
  date: string;
  status: TransactionStatus;
}

function signTransaction(transaction: Transaction) {
  const sign = signData(JSON.stringify(transaction), transaction.sender);
  const signedTransaction: SignedTransaction = { transaction, sign };
  return signedTransaction;
}

export function sendTransaction(transaction: Transaction) {
  const signedTranscation: SignedTransaction = signTransaction(transaction);
  const hexTransaction = Buffer.from(
    JSON.stringify(signedTranscation.transaction),
    "utf8"
  ).toString("hex");
  const hexSign = signedTranscation.sign.toString("hex");
  const transformedTransaction: PayloadTransaction = {
    transaction: hexTransaction,
    sign: hexSign,
  };
  const payload = JSON.stringify(transformedTransaction);
  const response = post("/transaction", payload);
  return response;
}

export async function getAllUserTransaction(cpf: string) {
  const url = `/transactions/${cpf}`;
  const transactions: FetchedTransaction[] = await get(url);
  const mappedTransactions: MappedTransaction[] = transactions.map((transaction) => {
    return {
      transactionId: transaction.id,
      senderCpf: transaction.cpf_sender,
      senderName: transaction.name_sender,
      recipientCpf: transaction.cpf_recipient,
      recipientName: transaction.name_recipient,
      amount: transaction.amount,
      date: format(parseISO(transaction.event_datetime), "dd/MM/yyyy HH:mm:ss"),
      status: transaction.status,
    };
  });
  return mappedTransactions;
}
